cliquet/storage/postgresql/__init__.py

```diff
@@ -0,0 +1,458 @@
+        try:
+            with self.connect() as cursor:
+                cursor.execute("SELECT now();")
+            return True
+        except psycopg2.Error:
+            return False
+
+    def collection_timestamp(self, resource, user_id):
+        query = """
+        SELECT resource_timestamp(%(user_id)s, %(resource_name)s)::BIGINT
+            AS timestamp;
+        """
+        resource_name = classname(resource)
+        placeholders = dict(user_id=user_id, resource_name=resource_name)
+        with self.connect() as cursor:
+            cursor.execute(query, placeholders)
+            result = cursor.fetchone()
+        return result['timestamp']
+
+    def create(self, resource, user_id, record):
+        query = """
+        INSERT INTO records (user_id, resource_name, data)
+        VALUES (%(user_id)s, %(resource_name)s, %(data)s::json)
+        RETURNING id, last_modified::BIGINT
+        """
+        resource_name = classname(resource)
+        placeholders = dict(user_id=user_id,
+                            resource_name=resource_name,
+                            data=json.dumps(record))
+
+        with self.connect() as cursor:
+            self.check_unicity(resource, user_id, record)
+
+            cursor.execute(query, placeholders)
+            inserted = cursor.fetchone()
+
+        record = record.copy()
+        record[resource.id_field] = inserted['id']
+        record[resource.modified_field] = inserted['last_modified']
+        return record
+
+    def get(self, resource, user_id, record_id):
+        query = """
+        SELECT last_modified::BIGINT, data
+          FROM records
+         WHERE id = %(record_id)s
+        """
+        placeholders = dict(record_id=record_id)
+        with self.connect() as cursor:
+            cursor.execute(query, placeholders)
+            if cursor.rowcount == 0:
+                raise exceptions.RecordNotFoundError(record_id)
+            else:
+                result = cursor.fetchone()
+
+        record = result['data']
+        record[resource.id_field] = record_id
+        record[resource.modified_field] = result['last_modified']
+        return record
+
+    def update(self, resource, user_id, record_id, record):
+        query_create = """
+        INSERT INTO records (id, user_id, resource_name, data)
+        VALUES (%(record_id)s, %(user_id)s,
+                %(resource_name)s, %(data)s::json)
+        RETURNING last_modified::BIGINT
+        """
+
+        query_update = """
+        UPDATE records SET data=%(data)s::json
+        WHERE id = %(record_id)s
+        RETURNING last_modified::BIGINT
+        """
+        resource_name = classname(resource)
+        placeholders = dict(record_id=record_id,
+                            user_id=user_id,
+                            resource_name=resource_name,
+                            data=json.dumps(record))
+
+        with self.connect() as cursor:
+            self.check_unicity(resource, user_id, record)
```

Comment:
Oh, so we need to actually do a GET before updating (or creating) the value.

I suppose this means it's not possible to ad constraints on JSON fields.